addons/website/models/website.py

```diff
@@ -915,28 +915,49 @@ def _get_alternate_languages(self, canonical_params):
 
     @api.model
     def get_current_website(self, fallback=True):
-        if request and request.session.get('force_website_id'):
+        """ The current website is returned in the following order:
+        - (frontend only) the website forced in session `force_website_id`
+        - the website set in context
+        - (frontend only) the website matching the request's "domain"
+        - arbitrary the first website found in the database if `fallback` is set
+          to `True`
+        - empty browse record
+        """
+        is_frontend_request = request and getattr(request, 'is_frontend', False)
+        if is_frontend_request and request.session.get('force_website_id'):
             website_id = self.browse(request.session['force_website_id']).exists()
             if not website_id:
                 # Don't crash is session website got deleted
                 request.session.pop('force_website_id')
             else:
                 return website_id
 
-        website_id = self.env.context.get('website_id')
+        website_id = self.env.context.get('website_id', False)
         if website_id:
             return self.browse(website_id)
 
         if not request and not fallback:
             return self.browse(False)
 
-        # The format of `httprequest.host` is `domain:port`
-        domain_name = request and request.httprequest.host or ''
+        if is_frontend_request:
+            # The format of `httprequest.host` is `domain:port`
+            domain_name = request and request.httprequest.host or ''
+            website_id = self._get_current_website_id(domain_name, fallback=fallback)
+        elif fallback:
+            website_id = self._get_first_website_id()
```

Comment:
Probably better here to call `_get_current_website_id` if we have request but not frontend_request -> Instead of returning the random first one we find, try to get the one matching the request host instead.
Ideally we will get rid of fallback in master imo, will see.
Still trying to figure why are 3 tests failing